[MachineLICM] Allow hoisting loads from invariant address

igogo-x86 opened the pull request:
Sometimes, loads can appear in a loop after the LICM pass is executed
the final time. For example, ExpandMemCmp pass creates loads in a loop,
and one of the operands may be an invariant address.
This patch extends the pre-regalloc stage MachineLICM by allowing to
hoist invariant loads from loops that don't have any stores or calls
and allows load reorderings.

igogo-x86 committed:
Pre-commit tests

igogo-x86 committed:
[MachineLICM] Allow hoisting loads from invariant address

Sometimes, loads can appear in a loop after the LICM pass is executed
the final time. For example, ExpandMemCmp pass creates loads in a loop,
and one of the operands may be an invariant address.
This patch extends the pre-regalloc stage MachineLICM by allowing to
hoist invariant loads from loops that don't have any stores or calls
and allows load reorderings.

igogo-x86 commented:
This patch affects several old tests, and I processed them differently:

1) Regression tests. I just updated the test lines:
* RISCV/rvv/fold-scalar-load-crash.ll
* X86/2009-04-25-CoalescerBug.ll
* X86/pr49393.ll
* X86/pr53842.ll

2) I need advice on what to do with these tests. They are testing some functionality but don't care about loads. I see three options - add a flag (as I did in my initial commit), update test lines as I did for regression tests or update tests by hoisting loads outside of loops on the IR level in a separate patch:

* X86/fma-commute-loop.ll
* AArch64/sinksplat.ll
* AArch64/ragreedy-local-interval-cost.ll
* AArch64/sinksplat.ll
* AArch64/zext-to-tbl.ll

3) The Test I don't understand, so I just updated the test line:
* X86/block-placement.ll

igogo-x86 committed:
Remove rudiment and apply formating

david-arm commented on llvm/test/CodeGen/AArch64/sinksplat.ll:
```diff
@@ -1,5 +1,5 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
-; RUN: llc -mtriple=aarch64-linux-gnu -o - %s | FileCheck %s
+; RUN: llc -mtriple=aarch64-linux-gnu -hoist-const-loads=false -o - %s | FileCheck %s
```
THIS IS FOR LINE 12: It doesn't look like the load is critical to the test - I think we care about sinking the splat into the loop so that we can use the right form of smlal (i.e. the indexed variant). So perhaps you can just update the test with the load hoisted out?

igogo-x86 replied:
Do you mean to hoist the load manually in a separate commit or just update the check lines together with the patch?

david-arm replied:
Sorry, what I said was confusing! I mean the latter - just recreate the CHECK lines with the patch without adding `-hoist-const-loads=false`. So long as the remaining instructions in the loop look the same as before I think it should be valid.

david-arm commented on llvm/test/CodeGen/AArch64/machine-licm-hoist-load.ll:
```diff
@@ -0,0 +1,428 @@
+; CHECK-NEXT:    mov x0, x8
+; CHECK-NEXT:    ret
+entry:
+  %cmp4 = icmp eq i64 %N, 0
```
I think you can simplify all the tests by removing the comparison of N, M, etc. with 0. We can just unconditionally jump to the `%for.body` block.

igogo-x86 replied:
Yeah, I missed this moment. Turns out there are a lot of places where I can make these tests smaller

david-arm commented on llvm/test/CodeGen/AArch64/machine-licm-hoist-load.ll:
```diff
@@ -0,0 +1,428 @@
+  %cmp4 = icmp eq i64 %N, 0
+  br i1 %cmp4, label %for.cond.cleanup, label %for.body
+
+for.cond.cleanup:                                 ; preds = %for.body, %entry
```
It would be nice to have the return blocks all at the end of the function.

david-arm commented on llvm/test/CodeGen/AArch64/machine-licm-hoist-load.ll:
```diff
@@ -0,0 +1,428 @@
+  ret i64 %sum.0.lcssa
+}
+
+define i64 @three_dimensional_middle(ptr %a, ptr %b, i64 %N, i64 %M, i64 %K) {
```
Might be useful to have a comment explaining how this is different to `@three_dimensional`, since it's not immediately obvious. Or you could even have simple C equivalents in comments above the functions?

david-arm commented on llvm/test/CodeGen/AArch64/machine-licm-hoist-load.ll:
```diff
@@ -0,0 +1,428 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc -mtriple=aarch64-linux-gnu < %s | FileCheck %s
+
+define i64 @one_dimensional(ptr %a, ptr %b, i64 %N, i64 %M, i64 %K) {
```
It would also be nice to have a version of this with memcmp of 5 or 6 bytes to show that both load (+ rev + or) instructions get hoisted out.

david-arm reviewed (commented):
Thanks for this @igogo-x86! Looks like a nice improvement, and is in line with what we already do for hoisting stores. I've not reviewed the code changes, but left some comments on the tests.

igogo-x86 committed:
Address the comments and fix failing tests

AArch64/machine-licm-hoist-load.ll - added memcmp-sized 6 test, comments, removed irrelevant lines and from AArch64/machine-licm-hoist-load.ll

AArch64/sinksplat.ll, AArch64/zext-to-tbl.ll - regenerated check lines without `-hoist-const-loads=false`

Adjusted tests after seeing that manually hoisted load results in the same assembly as the code where MachineLICM does this job:
* Hexagon/reg-scavengebug-2.ll
* Mips/lcb5.ll
* Hexagon/swp-const-tc2.ll (added `-hoist-const-loads=false` flag as I am unsure about the test's purpose)

igogo-x86 commented on llvm/test/CodeGen/RISCV/rvv/fold-scalar-load-crash.ll:
This regression test has guarded against some crashes in the past, so it is probably better to keep the set of flags closer to the default.

david-arm reviewed (commented):
Seems like a useful improvement to have @igogo-x86! I'd expect the IR version of LICM to hoist out most cases, but things like ExpandMemCmp, CodeGenPrepare or other pre-isel passes may introduce loads in loops that we do want to hoist out where possible.

igogo-x86 committed:
Refactor code

igogo-x86 commented on llvm/lib/CodeGen/MachineLICM.cpp:
```diff
@@ -772,6 +779,32 @@ void MachineLICMBase::HoistOutOfLoop(MachineDomTreeNode *HeaderN,
 
+  // Compute information about whether it is allowed to move load instruction
+  // out of the current loop or one of the inner loops
+  SmallDenseMap<MachineLoop *, bool> AllowedToHoistLoads;
```
Good idea! Done

RKSimon commented on llvm/test/CodeGen/X86/fma-commute-loop.ll:
```diff
@@ -1,5 +1,5 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
-; RUN: llc < %s -mtriple=x86_64-apple-darwin -mattr=avx512f | FileCheck %s
+; RUN: llc < %s -mtriple=x86_64-apple-darwin -mattr=avx512f -hoist-const-loads=false | FileCheck %s
```
Please can you drop this so we can see the test diff?

igogo-x86 committed:
Update fma-commute-loop.ll after removing flag

david-arm commented on llvm/lib/CodeGen/MachineLICM.cpp:
```diff
@@ -369,6 +377,36 @@ bool MachineLICMBase::runOnMachineFunction(MachineFunction &MF) {
+    while (!TmpWorklist.empty()) {
+      auto *L = TmpWorklist.pop_back_val();
+      AllowedToHoistLoads[L] = true;
+      TmpWorklist.insert(TmpWorklist.end(), L->getSubLoops().begin(),
```
Oh, does MLI not already include subloops? I was hoping you could get away with just:

```
for (auto *L : Worklist) {
  AllowedToHoistLoads[L] = true;
}
```

It's quite likely that was a bogus assumption. :)

igogo-x86 replied:
Yeah, at least it is the same behaviour as with IR Loops

david-arm commented on llvm/lib/CodeGen/MachineLICM.cpp:
```diff
@@ -369,6 +377,36 @@ bool MachineLICMBase::runOnMachineFunction(MachineFunction &MF) {
+    // Go through all the instructions inside top-level loops and, after finding
+    // one that makes it potentially unsafe to move loads, update load hoisting
+    // information for each loop containing this instruction.
+    for (auto *TopLoop : Worklist) {
```
Is it worth moving this loop into a small member function, i.e. something like `determineHoistableLoadLoops`? Not sure that's the best name though!

david-arm reviewed (commented):
This looks a lot neater now @igogo-x86, thanks! I just have two more minor comments, then it looks good I think.

igogo-x86 committed:
Refatctor initialization to a separate function

Also improve the algorithm to early exit when loop chain is
proven to be unhoistable

david-arm reviewed (approved):
LGTM! Thanks for all the changes @igogo-x86.

fhahn commented on llvm/lib/CodeGen/MachineLICM.cpp:
```diff
@@ -1333,6 +1346,46 @@ void MachineLICMBase::InitCSEMap(MachineBasicBlock *BB) {
+      if (!AllowedToHoistLoads[Loop])
+        continue;
+      for (auto &MI : *MBB) {
+        if (!MI.mayStore() && !MI.isCall() &&
```
It looks like this doesn't consider instructions with  unmodelled sideeffects, like memory barriers and we now (incorrectly) hoist loads across memory barriers. I put up https://github.com/llvm/llvm-project/pull/116987
